### PR TITLE
MAINT: Use scipy builtins rotation to compute quaternions.

### PR DESCRIPTION
**Pull request by Carreau**
As often, it is better to rely on existing battle tested code than roll our own. This also help testing the upstream code more.

This replace our own quaternion to euler angles with using the Rotations call in scipy.

This file comes from far at the beginning of napari, and had one significant modification in #2530 (WRT degenerate case), which I believe scipy already handle with the gimbals lock warning and is tested.

Then α, β, γ is not unique when β=±π/2, then current new implementation would have `α=<some value>`, `γ=0`, the old code `α=0`, `γ=2.π-<some value>`.

Note that the outputs are slightly different, with sometime the angles being +/- pi , +/- zero i in some case.

You can test the first commit of this branch with this test script that has a few hardcoded degenerate case, as well as pick ~10k random quaternion and compare the resutls.

https://gist.github.com/Carreau/53dfaefa0c07aa3a7a0bd90e890810bb

```
+/- zero
[-2.86, -3.12, -2.88, -2.64]
   old  1.491, 0.000, 1.658
   new  1.491, -0.000, 1.658
   diff 0.000, 0.000, -0.000
```

```
 +/- pi
[1.44, 0.48, 3.93, -1.31]
   old  -3.142, 0.702, -2.498
   new  3.142, 0.702, -2.498
   diff 6.283, 0.000, 0.000
```

```
gimbal lock:

[0.09, -0.78, 0.09, 0.78]
   old  0.000, 1.571, -2.912
   new  2.912, 1.571, 0.000
   diff 2.912, 0.000, -2.912

[-0.64, -0.58, -0.64, 0.58]
   old  0.000, 1.571, 1.473
   new  4.811, 1.571, 0.000
   diff 4.811, 0.000, 1.473
```

### CODE REVIEW

**Comment by Carreau** [comment]
Opening as draft, if we think we are ok with this I can remove completely the old function and rework the documentation, but I wanted to make sure I don't miss anything.

**Comment by Carreau** [comment]
Added a commit that remove old impl and cleanup docs.

**Comment by Carreau** [comment]
as a note, this function is _always_ used with `degree=True`, I would be inclined to remove the `degree` keyword and always assume it's True.

**Review by brisvag — APPROVED** [review]
Nice! I'm in favour of removing the kwarg since it's unneeded.

File: napari/_vispy/utils/quaternion.py
Comment (Czaki) [inline napari/_vispy/utils/quaternion.py:25]: Each time the `warnings.catch_warnings()` is used then a list of already seen warnings is reset. 
So using it in the context of often-used functions may result in bigger spamming of user. Why warning happen and how we avoid it?

File: napari/_vispy/utils/quaternion.py
Comment (Czaki) [inline napari/_vispy/utils/quaternion.py]:
```suggestion
    return x, y, z
```

File: napari/_vispy/utils/quaternion.py
Comment (Carreau) [reply napari/_vispy/utils/quaternion.py:25]: We can't avoid it because [math](https://en.wikipedia.org/wiki/Gimbal_lock), it is in SciPy and is triggered when the set of Euler Angles are degenerate and you are looking directly up a pole.

Can you point me to the issue or an explanation of "catch warnings" reseting the warning list ?

Comment (Czaki) [reply napari/_vispy/utils/quaternion.py:25]: https://github.com/python/cpython/issues/73858

Comment (Carreau) [reply napari/_vispy/utils/quaternion.py:25]: I could send a pr to SciPy with a parameter `gimbal_lock_ok:bool` to prevent the warning.

**Comment by jni** [comment]
@Carreau @brisvag personally I would keep `degree=True` (ie revert f9784a4) because as someone reading the code I am never sure of which units the angles are/should be in, and this documents that clearly.

Other than that I am +1 on these changes, minus the caveat from @Czaki — which I suppose would take some time, plus we try to follow SPEC0 for SciPy so it'll be like 2y? How do we handle this? Shall we fix that CPython issue? Wait that'll take longer... 😂

**Comment by Carreau** [comment]
> I would keep `degree=True` (ie revert [f9784a4](https://github.com/napari/napari/commit/f9784a437e1b2c0e1a2bb62e2d25dd3a04561a2e)) because as someone reading the code I am never sure of which units the angles are/should be in, and this documents that clearly.

That's fair.

> Shall we fix that CPython issue

There is already [a 5 Years Old PR](https://github.com/python/cpython/pull/8232) that received no comments, so unless we can poke at some CPython core dev to get that fixed...

Another possibility is to insert the filter once and for all.

**Comment by Carreau** [comment]
> I would keep `degree=True` (ie revert [f9784a4](https://github.com/napari/napari/commit/f9784a437e1b2c0e1a2bb62e2d25dd3a04561a2e)) because as someone reading the code I am never sure of which units the angles are/should be in, and this documents that clearly.

Would you be ok or prefer to rename to `quaternion2euler_degrees` or similar ? This way we don't have the keyword, and it is explicit. I just don't like having an unused keyword and adding types to a function which types depends on a boolean is possible but annoying.

**Comment by Czaki** [comment]
> Other than that I am +1 on these changes, minus the caveat from @Czaki — which I suppose would take some time, plus we try to follow SPEC0 for SciPy so it'll be like 2y? How do we handle this? Shall we fix that CPython issue? Wait that'll take longer... 😂

I do not feel that it is important to not emit warnings on non latest Scipy version.

**Comment by jni** [comment]
@Carreau yes I'm ok with renaming the function to include "degrees". 🙏

**Comment by brisvag** [comment]
@Carreau I didn't realize this was still marked as ready to merge despite the issue with catch-warnings pointed out by @Czaki, so I shouldn't have merged. I'll revert the PR. Cna you then reopen it while we wait for scipy to merge your chnages?

**Comment by brisvag** [comment]
For reference: https://github.com/scipy/scipy/issues/19306

**Comment by Carreau** [comment]
Do you want to reset hard main to scipy/spatial/transform/tests/test_rotation.py , and I can re-submit cc8d3ec9c4ecc0d461382688a10097b1dc7ce5e2 ? This way we don't have a revert on Main ?

**Comment by brisvag** [comment]
As mentioned on zulip, no problem with reverting with a new PR, but more pain comes from rewriting history. Let's move on with a new Pr!

**Comment by Carreau** [comment]
> As mentioned on zulip, no problem with reverting with a new PR, but more pain comes from rewriting history. Let's move on with a new Pr!

No worries, I'll reopen a PR once I have some confirmation that SciPy has the featuer upstream

**Pull request merged**
